src/coreclr/gc/objecthandle.cpp

```diff
@@ -1529,6 +1533,172 @@ void Ref_ScanSizedRefHandles(uint32_t condemned, uint32_t maxgen, ScanContext* s
 }
 #endif // FEATURE_SIZED_REF_HANDLES
 
+#ifdef FEATURE_GCBRIDGE
+
+// A trivial list data strucuture to hold the cross-reference entries
+class CrossReferenceList final
+{
+private:
+    struct Node
+    {
+        _UNCHECKED_OBJECTREF* pObjRef;
+        uintptr_t pExtraInfo;
+        Node* next;
+    };
+
+    Node* head;
+    size_t count;
+
+public:
+    CrossReferenceList() : head(nullptr), count(0) {}
+
+    ~CrossReferenceList()
+    {
+        Node* current = head;
+        while (current != nullptr)
+        {
+            Node* next = current->next;
+            delete current;
+            current = next;
+        }
+    }
+
+    void Add(_UNCHECKED_OBJECTREF* pObjRef, uintptr_t pExtraInfo)
+    {
+        Node* newNode = new Node{ pObjRef, pExtraInfo, head };
+        head = newNode;
+        ++count;
+    }
+
+    size_t Count() const
+    {
+        return count;
+    }
+
+    class Iterator
+    {
+    private:
+        Node* current;
+
+    public:
+        Iterator(Node* start) : current(start) {}
+
+        bool operator!=(const Iterator& other) const
+        {
+            return current != other.current;
+        }
+
+        void operator++()
+        {
+            if (current != nullptr)
+                current = current->next;
+        }
+
+        std::pair<_UNCHECKED_OBJECTREF*, uintptr_t> operator*() const
+        {
+            return { current->pObjRef, current->pExtraInfo };
+        }
+    };
+
+    Iterator begin() const
+    {
+        return Iterator(head);
+    }
+
+    Iterator end() const
+    {
+        return Iterator(nullptr);
+    }
+};
+
+struct Ops
+{
+    ScanContext* sc;
+    Ref_promote_func* pfn;
+};
+
+void CALLBACK CollectCrossReferenceEntries(_UNCHECKED_OBJECTREF *pObjRef, uintptr_t *pExtraInfo, uintptr_t lp1, uintptr_t lp2)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(pObjRef != NULL);
+    _ASSERTE(lp1 != NULL);
+    _ASSERTE(lp2 != NULL);
+
+    Object *pObj = VolatileLoad((PTR_Object*)pObjRef);
+    _ASSERTE(pObj != NULL);
+
+    // The object is already marked, so we don't need to ask the bridge about it.
+    if (g_theGCHeap->IsPromoted(pObj))
+        return;
+
+    // Promote the object and let the bridge indicate when the
+    // object is actually dead.
+    Ops* pOps = (Ops*)lp1;
+    pOps->pfn(&pObj, pOps->sc, 0);
```

Comment:
I do not think we want to be promoting when collecting the list of potentially dead objects.

Promoting an object can end up promoting other objects recursively. Some of these other objects may be participating in the GC bridge as well, and so we would miss them when building the graph.

Reply:
Promoting the objects should be a separate pass.

Reply:
Yes, this needs more investigation. I need the GC team or more time to learn a better approach in this area.